DREAM/Optimization/tsgParticleSwarm.cpp

```diff
@@ -0,0 +1,146 @@
+            int idx = i * num_dimensions + j;
+            particle_positions[idx] = range * get_random01() + box_lower[j];
+            particle_velocities[idx] = 2 * range * get_random01() - range;
+        }
+    }
+    initialized = true;
+}
+
+
+void ParticleSwarm(ObjectiveFunction f, int max_iterations, TasDREAM::DreamDomain inside, ParticleSwarmState &state,
+                   double inertia_weight, double cognitive_coeff, double social_coeff, std::function<double(void)> get_random01) {
+
+    // Only run the algorithm on properly initialized states.
+    if (!state.initialized) {
+        throw std::runtime_error("Particles have not been initialized in the input state object");
+    }
+
+    // Initialize helper variables and functions.
+    size_t num_dimensions = (size_t) state.getNumDimensions();
+    size_t num_particles = (size_t) state.getNumParticles();
+    std::vector<double> &particle_positions = state.getParticlePositionsRef();
+    std::vector<double> &particle_velocities = state.getParticleVelocitiesRef();
+    std::vector<double> &best_particle_positions = state.getBestParticlePositionsRef();
+    std::vector<double> &best_position = state.getBestPositionRef();
+    std::vector<double> prev_best_vals(num_particles), cur_best_vals(num_particles);
+    std::vector<bool> prev_is_inside(num_particles), cur_is_inside(num_particles);
+    double best_fval = std::numeric_limits<double>::max();
+    ObjectiveFunctionConstrained f_constrained = makeObjectiveFunctionConstrained(num_dimensions, f, inside);
+
+    // Create a lambda that updates the best particle and swarm positions based on function values and domain information.
+    auto update_best_positions = [&](const std::vector<double> prev_vals, const std::vector<double> cur_vals,
+                                     const std::vector<bool> prev_domain, const std::vector<bool> cur_domain)->void {
+        for (size_t i=0; i<num_particles; i++) {
+            bool is_smaller = (cur_vals[i] < prev_vals[i]) || !prev_domain[i];
+            if (cur_domain[i] && is_smaller) {
+                std::copy(particle_positions.begin() + i * num_dimensions, particle_positions.begin() + (i + 1) * num_dimensions,
+                          best_particle_positions.begin() + i * num_dimensions);
+                if (cur_vals[i] < best_fval) {
+                    std::copy(particle_positions.begin() + i * num_dimensions, particle_positions.begin() + (i + 1) * num_dimensions,
+                              best_position.begin());
+                    best_fval = cur_vals[i];
+                }
+            }
+        }
+    };
+
+    // Initialize the best particle and swarm function values, while updating the input state.
+    f_constrained(best_particle_positions, prev_best_vals, prev_is_inside);
+    f_constrained(particle_positions, cur_best_vals, cur_is_inside);
+    update_best_positions(prev_best_vals, cur_best_vals, prev_is_inside, cur_is_inside);
+    prev_best_vals = cur_best_vals;
```

Comment:
This is a bit redundant:
* let's just have "cur_vals" and "best_vals" and we overwrite "best_vals" if the corresponding "cur_val" is better, otherwise "cur_best" and "prev_best" is messy, after all best is best.
* if you want to make an update like this (and I don't think you should from the comment above), but if you do, then use `std::swap()` to avoid the full copy.